enhancement(http sink): improve config UX

KowalczykBartek opened the pull request:
Fix for https://github.com/vectordotdev/vector/pull/21583

KowalczykBartek commented:
@pront what do you think about this https://github.com/vectordotdev/vector/pull/21857/commits/bc3bc2863f758e3ad51240898c376ab680ad5107 ?

pront commented:
> @pront what do you think about this [bc3bc28](https://github.com/vectordotdev/vector/commit/bc3bc2863f758e3ad51240898c376ab680ad5107) ? this tls makes sense only for oauth2

Hi @KowalczykBartek, I appreciate you jumping on this so fast!
 
I am going to push a few commits to this PR and then I would really appreciate if you can run some manual tests again and share the config you used in this PR.

KowalczykBartek commented:
@pront you removed 
```rust
pub tls: Option<TlsConfig>,
```
now, there is no option to pass certs for mtls scenario

pront commented:
> @pront you removed
> 
> ```rust
> pub tls: Option<TlsConfig>,
> ```
> 
> now, there is no option to pass certs for mtls scenario - that was the idea behind this ticket, your http rest client can use different tls settings than cert/key used to acquire bearer token 

Ah yes indeed, feel free to add this back. Please just push commits directly to this PR and I will take a look. 

If you add that back, you can use `make generate-component-docs` to regenerate the docs.

You can also do `cd website && make run-production-site-locally` to see the result after the CUE files are updated. But that might need Hugo version 0.84.0. 
 
> > share the config you used in this PR.
> 
> [#21583 (comment)](https://github.com/vectordotdev/vector/pull/21583#issuecomment-2433783674)

Apologies if I wasn't clear, I meant that after this PR is finalized the config schema will change e.g. the `http_client_authorization_strategy` field should be removed right?

KowalczykBartek commented:
```
 feel free to add this back
```
but now there is a problem I wrote already : https://github.com/vectordotdev/vector/pull/21857#issuecomment-2489802458 
I don't know how to fix this.

pront commented:
> ```
>  feel free to add this back
> ```
> 
> but now there is a problem I wrote already : [#21857 (comment)](https://github.com/vectordotdev/vector/pull/21857#issuecomment-2489802458) I don't know how to fix this.

I see, it's because `TlsConfig` doesn't derive `PartialEq` and `Eq`. But you can just add them and everything should just work:

```
/// TLS configuration.
#[configurable_component]
#[configurable(metadata(docs::advanced))]
#[derive(Clone, Debug, Default, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct TlsConfig {
```

KowalczykBartek commented:
Ok, i think we need to revert this from master. This will just not work - https://github.com/vectordotdev/vector/pull/21857/commits/da06b02e2f1c5eb2d1bf7dd1def570b307cbf350 

```
sinks:
  my_sink_id:
    auth:
      strategy: "o_auth2"
      client_id: "xoxoxoxoxoxox"
      grace_period: 43100
      token_endpoint: "https://xoxoxoxoxox/oauth/token"
      tls:
        crt_path: xoxoxoxoxox/cert.pem
        key_file: xoxoxoxoxoxo/key.pem
    type: http
    encoding:
      codec: raw_message
    inputs:
      - my_source_id
    uri: http://localhost:9091/output
```
this will also kicks a 
```
    pub fn apply_headers_map(&self, map: &mut HeaderMap) {
        match &self {
            Auth::Basic { user, password } => {
                let auth = Authorization::basic(user.as_str(), password.inner());
                map.typed_insert(auth);
            }
            Auth::Bearer { token } => match Authorization::bearer(token.inner()) {
                Ok(auth) => map.typed_insert(auth),
                Err(error) => error!(message = "Invalid bearer token.", token = %token, %error),
            },
            Auth::OAuth2 { .. } => panic!("OAuth2 authentication is not supported currently"),
        }
    }
```
which just crash vector.

pront commented:
> Ok, i think we need to revert this from master. This will just not work - [da06b02](https://github.com/vectordotdev/vector/commit/da06b02e2f1c5eb2d1bf7dd1def570b307cbf350)
> 
> ```yaml
> sinks:
>   my_sink_id:
>     auth:
>       strategy: "o_auth2"
>       client_id: "xoxoxoxoxoxox"
>       grace_period: 43100
>       token_endpoint: "https://xoxoxoxoxox/oauth/token"
>       tls:
>         crt_path: xoxoxoxoxox/cert.pem
>         key_file: xoxoxoxoxoxo/key.pem
>     type: http
>     encoding:
>       codec: raw_message
>     inputs:
>       - my_source_id
>     uri: http://localhost:9091/output
> ```

The config format looks good 👍 
However, I cannot follow all the above. Are you suggesting to revert the original PR?

> this will also kicks a
> 
> ```rust
>     pub fn apply_headers_map(&self, map: &mut HeaderMap) {
>         match &self {
>             Auth::Basic { user, password } => {
>                 let auth = Authorization::basic(user.as_str(), password.inner());
>                 map.typed_insert(auth);
>             }
>             Auth::Bearer { token } => match Authorization::bearer(token.inner()) {
>                 Ok(auth) => map.typed_insert(auth),
>                 Err(error) => error!(message = "Invalid bearer token.", token = %token, %error),
>             },
>             Auth::OAuth2 { .. } => panic!("OAuth2 authentication is not supported currently"),
>         }
>     }
> ```
> 
> which just crash vector, that's why I created separate property, to distinguish that from the beginning
> 
> But I fixed a major refactoring, so at least its again rfc compliant :)

Does the RFC specify mandatory headers? If not, then we can replace the `panic!` with a `()` (noop).

KowalczykBartek commented:
```
Does the RFC specify mandatory headers? If not, then we can replace the panic! with a () (noop).
```
no no, both properties are now used by two different mechanisms, this conceptually is wrong, there needs to be different property than auth, cuz auth is already used

pront commented:
> @pront I propose to run this locally, with this config [#21857 (comment)](https://github.com/vectordotdev/vector/pull/21857#issuecomment-2492157529) (any oauth without tls atm is ok). then you will see the panic and why I am writing about two mechanism kicking in and what have to be done is explained here [#21583 (comment)](https://github.com/vectordotdev/vector/pull/21583#issuecomment-2442619253)

ACK. I doubt I can get to this before the release since I am occupied with some other work. Worst case, I will revert the original PR and then we can return to this without any time pressure.

KowalczykBartek commented:
In the worst case this PR will be a signpost for next brave hero :)

KowalczykBartek commented:
Ok, I got drunk yesterday (it's because of the weather, charms of eastern Europe..), and I had trouble reading, I went throught this discussion once again, so, lets summarize:

First, I think we were not on the same track, since this commit https://github.com/vectordotdev/vector/pull/21857/commits/bc3bc2863f758e3ad51240898c376ab680ad5107#diff-0820c0d03ae75b35f0b5bbd8dbc8e4d7ca048d19d60351bef069ac00c236910dL579 , @pront, you was referring to following state (here https://github.com/vectordotdev/vector/pull/21857#issuecomment-2492284570) :
```yaml
    # first
    auth:
      strategy: bearer
      token: xxx

   # second
   authorization_config:
      strategy: "o_auth2"
      client_id: "xoxoxoxoxoxox"
      grace_period: 43100
      token_endpoint: "https://xoxoxoxoxox/oauth/token"
      tls:
        crt_path: xoxoxoxoxox/cert.pem
        key_file: xoxoxoxoxoxo/key.pem

```
I was already assuming, there is need to remove authorization_config and rely only on 
```yaml
    auth:
      strategy: bearer
      token: xxx
```
(and, because of that, both mechism was kicked in when this auth was set, this lead to panic inside the vector)

I bring back this property https://github.com/vectordotdev/vector/pull/21857/commits/bf8f0f4094efa3d6b365dc081a7a029f9876973a#diff-0820c0d03ae75b35f0b5bbd8dbc8e4d7ca048d19d60351bef069ac00c236910dR575 so, right now, config looks like this 
```yaml
sinks:
  my_sink_id:
    authorization_config:
      strategy:
        strategy: "o_auth2"
        client_id: "myawesomeclientid"
        client_secret: "myawesomesecret"
        token_endpoint: "https://myawesomeoauthprovider/oauth/token"
        grace_period: 123
```
and, there is already existing Auth enum used, one problem (or not a problem) is that 
```yaml
      strategy:
        strategy: "o_auth2"
```
there is this duplicate strategy, but, I cannot make this exiting auth struct untagged, because its already tagged, it would infulance existing configuration 
```rust
#[configurable_component]
#[derive(Clone, Debug, Eq, PartialEq)]
#[serde(deny_unknown_fields, rename_all = "snake_case", tag = "strategy")]
#[configurable(metadata(docs::enum_tag_description = "The authentication strategy to use."))]
pub enum Auth {
```

(this commit included all steps from https://github.com/vectordotdev/vector/pull/21583#issuecomment-2450699285 )

pront commented:
> how can I build locally a dev docker image, I need to check something on k8s

I wonder if this `scripts/build-docker.sh` works in your environment. 
Otherwise, you can build locally and move the binary to a running docker instance.

---

Regarding the config UX, I will probably have to checkout this PR again and see if I can change the UX to just this:

```
    auth:
      strategy: "o_auth2"
      client_id: "xoxoxoxoxoxox"
      grace_period: 43100
      token_endpoint: "https://xoxoxoxoxox/oauth/token"
      tls:
        crt_path: xoxoxoxoxox/cert.pem
        key_file: xoxoxoxoxoxo/key.pem
```

From your message above it sounds like we don't need **two auth configs at the same time** but there are some details that get in the way. But bear with me, I am busy with other things so it might take a while.

KowalczykBartek commented:
> like we don't need two auth configs at the same time 

no, definitely. I would even say, that my solution is a replacement for existing(it does the same but its implemented in correct place, in the rest client), this method with apply_headers should is very (location is also strange).

> Regarding the config UX, I will probably have to checkout this PR again and see if I can change the UX to just this:

I don't think that's doable 

```diff
diff --git a/src/http.rs b/src/http.rs
index 4c849f1ef..dfa228fd6 100644
--- a/src/http.rs
+++ b/src/http.rs
@@ -580,6 +580,7 @@ impl<B> fmt::Debug for HttpClient<B> {
 pub struct AuthorizationConfig {
     /// Define how to authorize against an upstream.
     #[configurable]
+    #[serde(flatten)]
     strategy: Auth,
 
     /// The TLS settings for the http client's connection.
@@ -595,7 +596,7 @@ pub struct AuthorizationConfig {
 /// HTTP header without any additional encryption beyond what is provided by the transport itself.
 #[configurable_component]
 #[derive(Clone, Debug, Eq, PartialEq)]
-#[serde(deny_unknown_fields, rename_all = "snake_case", tag = "strategy")]
+#[serde(untagged, deny_unknown_fields)]
 #[configurable(metadata(docs::enum_tag_description = "The authentication strategy to use."))]
 pub enum Auth {
     /// Basic authentication.
```
but it doesn't work 
```
vector::cli: Configuration error. error=data did not match any variant of untagged enum Auth
```

estherk15 commented on website/cue/reference/components/sinks/base/clickhouse.cue:
```diff
@@ -36,6 +36,33 @@ base: components: sinks: clickhouse: configuration: {
+			}
+			grace_period: {
+				description: """
+					The grace period configuration for a bearer token.
+					To avoid random authorization failures caused by expired token exception,
+					we will acquire new token, some time (grace period) before current token will be expired,
+					because of that, we will always execute request with fresh enough token.
```
```suggestion
					The grace period configuration for a bearer token.
					This helps prevent random authorization failures due to expired tokens. It acquires a new token a short time before the current token expires, ensuring requests are always executed with a valid token.
```

estherk15 commented on website/cue/reference/components/sinks/base/clickhouse.cue:
```diff
@@ -58,6 +85,37 @@ base: components: sinks: clickhouse: configuration: {
+					o_auth2: """
+						Authentication based on OAuth 2.0 protocol.
+
+						This strategy allows to dynamically acquire and use token based on provided parameters.
+						Both standard client_credentials and mTLS extension is supported, for standard client_credentials just provide both
+						client_id and client_secret parameters:
```
```suggestion
						This strategy enables dynamic acquisition and usage of tokens based on the provided parameters.
						It supports both standard `client_credentials` and mTLS extensions. For standard `client_credentials`, provide both
						`client_id` and `client_secret` parameters.
```

estherk15 commented on website/cue/reference/components/sinks/base/clickhouse.cue:
````diff
@@ -58,6 +85,37 @@ base: components: sinks: clickhouse: configuration: {
+						 client_secret: "secret-value"
+						 token_endpoint: "https://yourendpoint.com/oauth/token"
+						```
+						In case you want to use mTLS extension [rfc8705](https://datatracker.ietf.org/doc/html/rfc8705), provide desired key and certificate,
+						together with client_id (with no client_secret parameter).
````
```suggestion
						To use mTLS extension [rfc8705](https://datatracker.ietf.org/doc/html/rfc8705), provide the desired key, certificate, and client_id (with no client_secret parameter).
```

pront commented:
Ok one last effort to sync on this. Otherwise, I am going to revert https://github.com/vectordotdev/vector/pull/21583 to unblock the next release. And hopefully we can tackle this in the future.

---

> my solution is a replacement for existing(it does the same but its implemented in correct place, in the rest client)

So we have:

1. basic (static user/pass creds)
2. brearer (static string) 
3. token provided by endpoint (dynamic)

Right?

Can you explain to me why we need to modify the existing code when we have (1) or (2)? 

I would expect that when you use (3) you can do whatever you want.

I don't see why we cannot add the 3rd one to this struct.

---

> this method with apply_headers is very limited (location is also strange).

This made me think....

I would expect that each HTTP request sets the OAuth token (as provided by the new endpoint. 
Example: 
```
POST /api/resource HTTP/1.1
Host: api.example.com
Authorization: Bearer abc123xyz
Content-Type: application/json

{
  "data": "example"
}
```
Yes/no?

KowalczykBartek commented:
```
Can you explain to me why we need to modify the existing code when we have (1) or (2)?
```
where I modified existing codebase ? I provided implementation strictly for oauth token acquisition, additionally I added for user+password (so, my implementation, can be drop-in-replacement for non async apply_headers method).

pront commented:
> additionally I added for user+password (so, my implementation, can be drop-in-replacement for non async apply_headers method).

Gotcha, that's where the confusion was. Since this is a critical component I am afraid we are doing way too much at once. I am open to replace the existing authentication method but as a two-phase process. So basically, on path forward here is:

* keep (1) and (2) as is (important)
  * could be deprecated in favor of the new one
  * but let's not advertise as deprecated yet
* also keep the new `AuthorizationConfig` as a new beta feature
  * document that it is a replacement for the existing `auth` field  
  * Introduce a `Auth2` enum to make it easier to maintain
  * rename `strategy: Auth,` to `auth: Auth2` 
    *   * let me come back to you, I will check locally if we can flatten it ;)   
* add config validation to throw an error if both are defined (important)
  * this basically prohibits setting both

Does this make sense to you?

KowalczykBartek commented:
```
i will go ahead and revert
```
pls go ahead, unfortunately i cannot spend more time on this

KowalczykBartek commented:
@pront 
```
but we really want to avoid breaking existing users.
```
break what ?? I provided 100% backward compatible implementation

pront commented:
> pls go ahead, unfortunately i cannot spend more time on this

Thank you for your time. In the future, we will revive https://github.com/vectordotdev/vector/pull/21583 and address the UX issue as outlined [here](https://github.com/vectordotdev/vector/pull/21857#issuecomment-2499083493).
